util/coverage/coverage_off_target_test.py

```diff
@@ -0,0 +1,237 @@
+        libs_all = ["//foo/bar1", "//foo/bar2", "//foo/incompat"]
+        libs_incompat = ["//foo/incompat"]
+        libs = list(set(libs_all) - set(libs_incompat))
+        # (1) return all libs, (2) return incompat., (3) build
+        mock_run.side_effect = (libs_all, libs_incompat, [])
+
+        self.assertEqual(coverage_off_target.gather_and_build_libs(), libs)
+
+        calls = mock_run.call_args_list
+        self.assertEqual(len(calls), 3)
+        self.assertEqual(calls[0][0][:2], (coverage_off_target.BAZEL, "query"))
+        self.assertEqual(calls[1][0][:2], (coverage_off_target.BAZEL, "cquery"))
+        self.assertEqual(calls[2][0], (coverage_off_target.BAZEL, "build",
+                                       "--config=coverage_clang", *libs))
+
+    @patch("coverage_off_target.run")
+    def test_create_merged_library(self, mock_run):
+        libs = ["//foo/bar1", "//foo/bar2"]
+        libs_query = "(//foo/bar1 + //foo/bar2)"
+        merged_so = Path("merged.so")
+        objs = ["foo/bar1.o", "foo/bar2.o"]
+        # (1) return object files, (2) link.
+        mock_run.side_effect = (objs, [])
+
+        coverage_off_target.create_merged_library(libs, merged_so)
+
+        calls = mock_run.call_args_list
+        self.assertEqual(len(calls), 2)
+        self.assertEqual(calls[0][0][:3],
+                         (coverage_off_target.BAZEL, "cquery", libs_query))
+        self.assertEqual(
+            calls[1][0],
+            (coverage_off_target.LLD, "--shared", "-o", str(merged_so), *objs))
+
+    @patch("coverage_off_target.run")
+    def test_measure_unit_test_coverage(self, mock_run):
+        tests = ["//foo:test1", "//foo:test2"]
+        testlogs = ["/some/long/path"]
+        profiles = [
+            "/some/long/path/foo/test1/coverage.dat",
+            "/some/long/path/foo/test2/coverage.dat"
+        ]
+        merged_profile = Path("merged.profdata")
+        # (1) find test targets, (2) measure coverage, (3) get testlogs dir, (4) merge
+        # profile data.
+        mock_run.side_effect = (tests, [], testlogs, [])
+
+        coverage_off_target.measure_unit_test_coverage(merged_profile)
+
+        calls = mock_run.call_args_list
+        self.assertEqual(len(calls), 4)
+        self.assertEqual(calls[0][0][:2], (coverage_off_target.BAZEL, "query"))
+        self.assertEqual(calls[1][0], (coverage_off_target.BAZEL, "coverage",
+                                       "--config=coverage_clang", *tests))
+        self.assertEqual(calls[2][0],
+                         (coverage_off_target.BAZEL, "info", "bazel-testlogs"))
+        self.assertEqual(
+            calls[3][0],
+            (coverage_off_target.LLVM_PROFDATA, "merge", "--sparse", "--output",
+             str(merged_profile), *profiles))
+
+    @patch("coverage_off_target.print")
+    @patch("pathlib.Path.open", new_callable=mock_open)
+    @patch("pathlib.Path.__truediv__",
+           side_effect=Path.__truediv__,
+           autospec=True)
```

Comment:
Could you add a comment explaining what this does?

It looks like the mock either forwards its arguments to the *real* implementation or it's a boundlessly recursive mock, but I figure both of those are incorrect.